src/Control/Retry.hs

```diff
@@ -246,29 +301,30 @@ capDelay limit p = RetryPolicyM $ \ n ->
 --
 -- Note how the latest failing result is returned after all retries
 -- have been exhausted.
-retrying :: MonadIO m
-         => RetryPolicyM m
-         -> (Int -> b -> m Bool)
-         -- ^ An action to check whether the result should be retried.
-         -- If True, we delay and retry the operation.
-         -> m b
-         -- ^ Action to run
-         -> m b
-retrying (RetryPolicyM policy) chk f = go 0
-    where
-      go n = do
-          res <- f
-          chk' <- chk n res
-          case chk' of
-            True -> do
-              chk <- policy n
-              case chk of
-                Just delay -> do
-                  liftIO (threadDelay delay)
-                  go $! n+1
-                Nothing -> return res
-            False -> return res
-
+retrying  :: MonadIO m
+          => RetryPolicyM m
+          -> (RetryStatus -> b -> m Bool)
+          -- ^ An action to check whether the result should be retried.
+          -- If True, we delay and retry the operation.
+          -> (RetryStatus -> m b)
+          -- ^ Action to run
+          -> m b
+retrying (RetryPolicyM policy) chk f = go defaultRetryStatus
+  where
+    go s = do
+        res <- f s
+        chk' <- chk s res
+        case chk' of
+          True -> do
+            chk <- policy s
+            case chk of
+              Just delay -> do
+                liftIO (threadDelay delay)
+                go $! RetryStatus { rsIterNumber = rsIterNumber s + 1
+                                  , rsCumulativeDelay = rsCumulativeDelay s + delay
+                                  , rsPreviousDelay = Just (maybe 0 (const delay) (rsPreviousDelay s))}
```

Comment:
I think these record updates accumulate thunks if the client code `f` doesn't touch them. I think it's better to make those fields strict.